### DISCUSSION TITLE
Growatt message catalogue

### DISCUSSION THREAD

**Discussion by PedroKTFC** [johanmeijer/grott#499] [General]
I really appreciate Grott and use it to drive some of my automations, so many thanks for putting this together. Ultimately, I would like to completely decouple my system from the Growatt servers (for security and reliability reasons). Is there a catalogue of all the messages the data logger and Growatt servers exchange available (apologies if there is and I've missed it)? I imagine it's a work in progress.

I've started running a version of Grott locally in a terminal window so I can easily and quickly make changes and see their impact. For example, I tried turning off the forwarding od messages between the Grott server and the data logger/Growatt servers. Not surprsingly, I never saw any data as (presumably), the data logger and Growatt servers have a conversation before the data logger sends its data.

**Comment by johanmeijer** [comment]
You can use grottserver as a replacement for the growatt server, see: https://github.com/johanmeijer/grott/wiki/Grottserver. 

Grottserver will be integrated in a next version with Grott,  for now this are 2 separate applications. 

More about grottserver and the experiences of others you can also read in discussion #98.

**Comment by PedroKTFC** [comment]
Thanks. I came across the wiki shortly after raising my query, doh! I look forward to the combined version as it'll probably be easier to run them like that in HA. Is there a message catalogue documented anywhere? I know these are reverse engineered.

**Comment by patcher-ms** [comment]
I was looking into it recently and while I haven’t found any definitive resource I came across a couple pages that give have at least some details:

- https://www.vromans.org/johan/software/sw_growatt_wifi_protocol.html
- https://github.com/aaronjbrown/PyGrowatt/blob/master/scripts/Growatt.lua
- https://www.mariosangiorgio.com/notes/growatt-messages/

There is still work to do before we know what all the messages are but at least that’s a start.

**Reply by johanmeijer** [reply]
@PedroKTFC @patcher-ms 
Please be aware that most of the encoding is already done in Grottserver. This is a working replacement for the growatt server. 
(the document van Johan Vromans was de basis for al this, but the protocol has changed a lot during time). 

See the wiki: https://github.com/johanmeijer/grott/wiki/Grottserver and discussion #98 for more information.

**Reply by PedroKTFC** [reply]
Thanks. I've managed to catch up and now realise things are far more advanced than I had appreciated from just the Grott HA add-on.
